### ISSUE TITLE
user constantly being blocked by csf but no reason logged

### ISSUE THREAD

**Issue by santo74** [omega8cc/boa#728]
I have a user who is constantly being blocked by csf,  but I can't find any reasons in the logfiles why this is happening.

This is what's being logged in lfd.log:

```
Jun  1 00:10:52 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 00:10:53 server01 lfd[14836]: csf (re)start completed
Jun  1 01:09:54 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 01:09:55 server01 lfd[14836]: csf (re)start completed
Jun  1 02:05:46 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 02:05:47 server01 lfd[14836]: csf (re)start completed
Jun  1 03:05:32 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 03:05:33 server01 lfd[14836]: csf (re)start completed
Jun  1 04:18:04 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 04:18:05 server01 lfd[14836]: csf (re)start completed
Jun  1 05:16:31 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 05:16:32 server01 lfd[14836]: csf (re)start completed
Jun  1 06:08:18 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 06:08:19 server01 lfd[14836]: csf (re)start completed
Jun  1 07:11:29 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 07:11:30 server01 lfd[14836]: csf (re)start completed
Jun  1 08:16:31 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 08:16:32 server01 lfd[14836]: csf (re)start completed
Jun  1 09:06:53 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 09:06:54 server01 lfd[14836]: csf (re)start completed
Jun  1 10:07:55 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 10:07:56 server01 lfd[14836]: csf (re)start completed
Jun  1 11:09:26 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 11:09:27 server01 lfd[14836]: csf (re)start completed
Jun  1 12:15:44 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 12:15:45 server01 lfd[14836]: csf (re)start completed
Jun  1 12:40:06 server01 lfd[32524]: Incoming IP <user-ip>:21 temporary block removed
Jun  1 12:40:06 server01 lfd[32524]: Incoming IP <user-ip>:22 temporary block removed
Jun  1 12:40:06 server01 lfd[32524]: Incoming IP <user-ip>:443 temporary block removed
Jun  1 12:40:06 server01 lfd[32524]: Incoming IP <user-ip>:80 temporary block removed
Jun  1 13:12:46 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 13:12:47 server01 lfd[14836]: csf (re)start completed
Jun  1 13:40:08 server01 lfd[8063]: Incoming IP <user-ip>:21 temporary block removed
Jun  1 13:40:08 server01 lfd[8063]: Incoming IP <user-ip>:22 temporary block removed
Jun  1 13:40:08 server01 lfd[8063]: Incoming IP <user-ip>:443 temporary block removed
Jun  1 13:40:08 server01 lfd[8063]: Incoming IP <user-ip>:80 temporary block removed
Jun  1 14:10:58 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 14:10:59 server01 lfd[14836]: csf (re)start completed
Jun  1 14:40:14 server01 lfd[15961]: Incoming IP <user-ip>:21 temporary block removed
Jun  1 14:40:14 server01 lfd[15961]: Incoming IP <user-ip>:22 temporary block removed
Jun  1 14:40:14 server01 lfd[15961]: Incoming IP <user-ip>:443 temporary block removed
Jun  1 14:40:14 server01 lfd[15961]: Incoming IP <user-ip>:80 temporary block removed
Jun  1 15:07:35 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 15:07:36 server01 lfd[14836]: csf (re)start completed
Jun  1 15:40:22 server01 lfd[24030]: Incoming IP <user-ip>:21 temporary block removed
Jun  1 15:40:22 server01 lfd[24030]: Incoming IP <user-ip>:22 temporary block removed
Jun  1 15:40:22 server01 lfd[24030]: Incoming IP <user-ip>:443 temporary block removed
Jun  1 15:40:22 server01 lfd[24030]: Incoming IP <user-ip>:80 temporary block removed
Jun  1 16:09:37 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 16:09:38 server01 lfd[14836]: csf (re)start completed
Jun  1 16:40:28 server01 lfd[29327]: Incoming IP <user-ip>:21 temporary block removed
Jun  1 16:40:28 server01 lfd[29327]: Incoming IP <user-ip>:22 temporary block removed
Jun  1 16:40:28 server01 lfd[29327]: Incoming IP <user-ip>:443 temporary block removed
Jun  1 16:40:28 server01 lfd[29327]: Incoming IP <user-ip>:80 temporary block removed
Jun  1 17:08:24 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 17:08:25 server01 lfd[14836]: csf (re)start completed
Jun  1 18:13:26 server01 lfd[14836]: csf (re)start requested - running *csf startup*...
Jun  1 18:13:27 server01 lfd[14836]: csf (re)start completed
```

**Comment by omega8cc** [comment]
Please check/grep the firewall logs for hints on what happens exactly:

`$ grep <user-ip> /var/xdrago/monitor/*.log`

**Comment by santo74** [comment]
Thanks, a grep revealed the following information:
```
cloud01:~# grep "<user-ip>" /var/xdrago/monitor/*.log
<user-ip> # [x255] 150601-113956
<user-ip> # [x255] 150601-121557
<user-ip> # [x255] 150601-131257
<user-ip> # [x255] 150601-141057
<user-ip> # [x255] 150601-150757
```
But what is this telling me about the cause of the blocks?
Does the [x255] represent the number of concurrent connections from that ip?

**Comment by omega8cc** [comment]
That is how DoS protection works. Please check:

https://omega8.cc/how-firewall-works-is-my-ip-blocked-121
https://github.com/omega8cc/boa/blob/master/docs/cnf/barracuda.cnf#L52

The user either crawls the site aggressively or has a redirect loops etc. You need to check /var/log/nginx/access.log for details.

**Comment by santo74** [comment]
ok, thanks for the pointers to the documentation.
So the "[x255]" does indicate the number of connections?

In any case I don't see anything too special in the nginx access logs and therefore still don't understand why the ip address would be blocked.

During the whole day, there were only 225 connections from that ip address:
```
server01:~# grep "<user-ip>" /var/log/nginx/access.log.2 | wc -l
225
```

Also, between 01/Jun/2015:11:39:49 and 01/Jun/2015:19:27:53 there wasn't any activiy in the access log from that ip address, while the lfd log mentions every hour between 12:40:06 and 16:40:28 for that day that the temporary block was removed for that ip address. (cfr my initial post).
In other words, after 12:40:06 the ip address got blocked again (several times) while there was no activity in the nginx access log.

So I guess there must be something else going on which is causing the blocks.

**Comment by omega8cc** [comment]
You may need to analyze also emails sent by lfd/csf. The user could run ping flood, port scans etc.

**Comment by santo74** [comment]
True, but the last email I received from lfd/csf (which related to port scanning by the way) dates from 14/04/2015 and is obviously not related to this issue.

**Comment by omega8cc** [comment]
If this was related to DoS protection, then it could happen only when no one else visited the site and the IP reached the limit 250/300. Please investigate all logs once this happens again. If you don't have enough "noise" in the access log, consider disabling the protection by using something like `_NGINX_DOS_LIMIT=900` and running `barracuda up-stable system`

**Comment by santo74** [comment]
Well, increasing the dos limit is of course a workaround but doesn't help me understand why csf/lfd is blocking the ip in the first place.
I would expect that lfd/csf always logs the reason somewhere in the logfiles and normally this is in the lfd.log file itself as far as I know, but apparently not on a boa system.

The fact that no details are logged in lfd.log, no abnormal activities are registered in the access logs and no emails are being sent on suspicious activities such as port scans makes it at least "strange" that the ip is being blocked all the time.

Anyway, I'll keep an eye on the logs but I don't expect any improvements in this regard.

**Comment by omega8cc** [comment]
DoS protection is a BOA feature, so there will be nothing logged by csf/lfd. You have identified the source of the problem in the DoS protection logs maintained by BOA. Now you need to examine the nginx access logs to determine what exactly caused that high rate of requests from a single IP address.

**Comment by omega8cc** [comment]
If the IP was blocked also when there is nothing in the DoS protection logs, then the block was probably caused by csf/lfd due to port scan or ping flood. That is everything we can guess.

**Comment by santo74** [comment]
Ok, I understand all that, but in any case I should have received more information, no?
In case of a DoS attack, an email should have been sent, as was the case with all previous DoS attacks on my systems.
In case of a port scan/ping flood, csf/lfd should have provided me with more details in its logfile.

**Comment by omega8cc** [comment]
No, no e-mails are expected to be sent on any such DoS event. You need to examine logs.
But when port scan or ping flood happens, then csf/lfd does send emails. Just make sure they don't land in your spam folder and/or examine /var/log/mail.log We don't modify csf/lfd.

**Comment by santo74** [comment]
I'm sorry, you're right that no emails are being sent in case of DoS events.
They are all about port scans, not DoS attacks. My mistake.

Unfortunately the problem remains that I really can't find any relation between the DoS protection logs and the nginx access logs.

But thanks for your support, I really appreciate it!

**Comment by Anoesj** [comment]
Just checking: does `NGINX_DOS_LIMIT=900` mean: "if 900 out of the last 300 (?) requests are from the same IP, add to CSF block"? Or does it mean "if 900 out of last 900 requests are from the same IP, add to CSF block".
Also, how about a `NGINX_DOS_ENABLED` flag? :smile:

**Comment by omega8cc** [comment]
You can enable/disable by modifying `NGINX_DOS_LIMIT `

Any value above 300 means it's disabled, because it will check only the last 300 lines anyway.
